### PR TITLE
Raise and Escalate Dispute Orchestration

### PR DESCRIPTION
**Pull request by cliffhall**
## What is it
An orchestration method that raises and escalates a dispute in a single transaction.

## Status
Method works, and all of the applicable tests in the raiseDispute and escalateDispute contexts of DisputeHandlerTest.js have been replicated in OrchestrationHandlerTest.js and made to work with the raiseAndEscalateDispute method of the OrchestrationHandler. 

One exception on the test coverage is that tests that were skipped in the DisputeHandlerTest.js file were omitted. They are forward looking tests that rely on the dispute resolver being able to set a non-zero fee. Thus any test of the native or ERC-20 payment of the escalation deposit can't be tested, because the deposit is a percentage of the fee, and the fee is always zero currently.

# What's changed
* In DisputeBase.sol,
  - import FundsLib
  - add disputesNotPaused modifier to raiseDisputeInternal which called from multiple places which do not use the modifier
  - added escalateDisputeInternal,
    - refactor/extracted from escalateDispute for reuse by orchestration

* In DisputeHandlerFacet.sol,
  - removed disputesNotPaused from raiseDispute method. raiseDisputeInternal will perform this check
  - updated natspec for escalateDispute

* In IBosonDisputeHandler.sol,
  - updated natspec for escalateDispute

* In IBosonOrchestrationHandle.sol,
  - added raiseAndEscalateDispute method

* In OrchestrationHandlerFacet.sol,
  - import and extend DisputeBase
  - added raiseAndEscalateDispute method

* In OrchestrationHandlerTest.js,
  - added imports 
    - setNextBlockTimestamp 
    - Dispute 
    - DisputeState
    - DisputeDates
  - added global vars for new tests
  - added wallets for new tests
  - grant PROTOCOL role to diamond for issuing vouchers
  - deployed facets for new tests
    - DisputeHandlerFacet
    - FundsHandlerFacet
  - removed beforeEach setting of validFrom and validUntil dates, which was using Date.now() which results in a shorter timestamp that isn't compatible with block.timestamp dates
  - fixed the one existing orchestration test that failed when the offer validity dates were changed back to the default dates set by the mock utility using block.timestamp
  - added context for raiseAndEscalateDispute
    - mostly a copy of the 'Single' context from DisputeHandlerFacetTest.js
    - beforeEach modified to work in the context of the seller/offer/etc that has already been set up globally in this test suite 
    - got tests running that validate most of the assertions from raiseDispute and escalateDispute tests combined. 
    - removed tests that were skipped in the DisputeHandlerTest.js suite that rely on the dispute resolver being able to set a non-zero fee which is impossible right now.

### CODE REVIEW

File: test/protocol/OrchestrationHandlerTest.js [hunk 1]
```diff
@@ -339,6 +366,278 @@ describe("IBosonOrchestrationHandler", function () {
...
+          // Attempt to raise a dispute, expecting revert
+          await expect(
+            orchestrationHandler
+              .connect(buyer)
+              .raiseAndEscalateDispute(exchangeId, { value: buyerEscalationDepositNative })
+          ).to.revertedWith(RevertReasons.REGION_PAUSED);
+        });
+
+        it("Caller is not the buyer for the given exchange id", async function () {
+          // Attempt to retract the dispute, expecting revert
```
Comment (zajck) [inline test/protocol/OrchestrationHandlerTest.js:566]:
```suggestion
          // Attempt to raise and escalate the dispute, expecting revert
```

File: contracts/interfaces/handlers/IBosonOrchestrationHandler.sol [hunk 2]
```diff
@@ -22,6 +22,32 @@ interface IBosonOrchestrationHandler is
...
     IBosonBundleEvents
 {
+    /**
+     * @notice Raises a dispute and immediately escalates it.
+     *
+     * Caller must send (or for ERC20, approve the transfer of) the
+     * buyer escalation deposit percentage of the offer price, which
+     * will be added to the pot for resolution.
+     *
+     * Emits a DisputeEscalated event if successful.
```
Comment (zajck) [inline contracts/interfaces/handlers/IBosonOrchestrationHandler.sol:32]:
```suggestion
     * Emits a DisputeRaised and a DisputeEscalated event if successful.
```

File: contracts/protocol/facets/OrchestrationHandlerFacet.sol [hunk 3]
```diff
@@ -33,6 +35,44 @@ contract OrchestrationHandlerFacet is
...
     }
 
+    /**
+     * @notice Raises a dispute and immediately escalates it.
+     *
+     * Caller must send (or for ERC20, approve the transfer of) the
+     * buyer escalation deposit percentage of the offer price, which
+     * will be added to the pot for resolution.
+     *
+     * Emits a DisputeEscalated event if successful.
```
Comment (zajck) [inline contracts/protocol/facets/OrchestrationHandlerFacet.sol:45]:
```suggestion
     * Emits a DisputeRaised and a DisputeEscalated event if successful.
```

**Review by zajck — CHANGES_REQUESTED** [review]
Looks good, I pointed out just few minor things.

And interfaceid in natspec in `IBosonOrchestrationHandler.sol` should be updated.

File: contracts/protocol/bases/DisputeBase.sol [hunk 4]
```diff
@@ -28,7 +29,7 @@ contract DisputeBase is ProtocolBase, IBosonDisputeEvents {
         Exchange storage _exchange,
```
Comment (anajuliabit) [inline contracts/protocol/bases/DisputeBase.sol:29]: Please add to `reverts if` section: 
```     * - The disputes region of protocol is paused```

File: contracts/protocol/facets/DisputeHandlerFacet.sol [hunk 5]
```diff
@@ -37,7 +37,7 @@ contract DisputeHandlerFacet is DisputeBase, IBosonDisputeHandler {
      *
```
Comment (anajuliabit) [inline contracts/protocol/facets/DisputeHandlerFacet.sol:37]: Please add to reverts if section:
- The disputes region of protocol is paused

File: contracts/protocol/facets/OrchestrationHandlerFacet.sol [hunk 6]
```diff
@@ -33,6 +35,44 @@ contract OrchestrationHandlerFacet is
...
+     *
+     * Reverts if:
+     * - The disputes region of protocol is paused
```
Comment (anajuliabit) [inline contracts/protocol/facets/OrchestrationHandlerFacet.sol:48]: Please add:
```suggestion
     * - The disputes region of protocol is paused
     * - The orchestration region of protocol is paused
```
```

**Review by anajuliabit — CHANGES_REQUESTED** [review]

File: contracts/protocol/bases/DisputeBase.sol (hunk 4 above)
Comment (cliffhall) [reply contracts/protocol/bases/DisputeBase.sol:29]: done

File: contracts/protocol/facets/DisputeHandlerFacet.sol (hunk 5 above)
Comment (cliffhall) [reply contracts/protocol/facets/DisputeHandlerFacet.sol:37]: done

Comment (anajuliabit) [reply contracts/protocol/facets/DisputeHandlerFacet.sol:38]: sorry for not being clearer, but it's missing from IBosonDisputeHandler.sol as well

File: contracts/protocol/facets/OrchestrationHandlerFacet.sol (hunk 6 above)
Comment (anajuliabit) [reply contracts/protocol/facets/OrchestrationHandlerFacet.sol:48]: sorry for not being clearer, but it's missing from IBosonOrchestationHandler.sol as well

**Review by mischat — COMMENTED** [review]
This looks good to me, i will just wait for the tests to finish running, but the code looks good

**Review by zajck — DISMISSED** [review]

**Review by anajuliabit — APPROVED** [review]

**Review by zajck — DISMISSED** [review]

**Review by anajuliabit — APPROVED** [review]
LGTM

**Review by mischat — APPROVED** [review]
This looks good to me, it seems nice and clean

**Pull request merged**